File: _streamlit_legacy/modules/mission_planner.py
"""Calcolo parametri di volo (GSD, quota, overlap) per missioni SAPR/SAR."""

from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path

DEFAULT_DRONES_DB = Path(__file__).resolve().parent.parent / "config" / "drones.json"

# Linee guida di overlap raccomandate per fotogrammetria/mappatura SAR
FRONT_OVERLAP_MIN, FRONT_OVERLAP_MAX = 70, 80
SIDE_OVERLAP_MIN, SIDE_OVERLAP_MAX = 60, 70

# Target GSD raccomandato per individuazione target SAR (dettaglio indumento/persona).
# Range verificato rispetto al dataset accademico Heridal (foto SAR scattate tra 40 e 65m,
# con accuratezza di rilevamento che cala fuori da quel range) e alle specifiche reali della
# flotta: per il M3E camera wide, 0.8-1.5cm corrisponde a circa 30-56m di quota.
GSD_TARGET_MIN_CM = 0.8
GSD_TARGET_MAX_CM = 1.5


@dataclass
class SensorProfile:
    id: str
    drone: str
    sensore: str
    sensor_width_mm: float | None
    sensor_height_mm: float | None
    image_width_px: int | None
    image_height_px: int | None
    focal_length_mm: float | None
    note: str = ""

    @property
    def is_custom(self) -> bool:
        return self.sensor_width_mm is None or self.focal_length_mm is None or self.image_width_px is None


def load_sensor_profiles(path: Path | str = DEFAULT_DRONES_DB) -> list[SensorProfile]:
    with open(path, encoding="utf-8") as f:
        data = json.load(f)
    return [SensorProfile(**entry) for entry in data["drones"]]


def gsd_cm_per_px(altitude_m: float, sensor_width_mm: float, focal_length_mm: float, image_width_px: int) -> float:
    """GSD (cm/pixel) in funzione della quota di volo (AGL, in metri)."""
    if focal_length_mm <= 0 or image_width_px <= 0:
        raise ValueError("focal_length_mm e image_width_px devono essere > 0")
    return (sensor_width_mm * altitude_m * 100) / (focal_length_mm * image_width_px)


def altitude_for_gsd(gsd_cm_px: float, sensor_width_mm: float, focal_length_mm: float, image_width_px: int) -> float:
    """Quota di volo (m AGL) necessaria per ottenere il GSD target."""
    if sensor_width_mm <= 0:
        raise ValueError("sensor_width_mm deve essere > 0")
    return (gsd_cm_px * focal_length_mm * image_width_px) / (sensor_width_mm * 100)


def recommended_altitude_range(
    sensor_width_mm: float,
    focal_length_mm: float,
    image_width_px: int,
    gsd_min_cm: float = GSD_TARGET_MIN_CM,
    gsd_max_cm: float = GSD_TARGET_MAX_CM,
) -> tuple[float, float]:
    """Range di quota (m AGL) che mantiene il GSD nell'intervallo target.

    Quota piu' bassa -> GSD piu' fine (gsd_min_cm); quota piu' alta -> GSD piu' grossolano (gsd_max_cm).
    """
    alt_min = altitude_for_gsd(gsd_min_cm, sensor_width_mm, focal_length_mm, image_width_px)
    alt_max = altitude_for_gsd(gsd_max_cm, sensor_width_mm, focal_length_mm, image_width_px)
    return alt_min, alt_max


def ground_footprint_m(altitude_m: float, sensor_width_mm: float, sensor_height_mm: float, focal_length_mm: float) -> tuple[float, float]:
    """Estensione a terra (larghezza, altezza in metri) inquadrata da un singolo fotogramma."""
    width_m = (sensor_width_mm * altitude_m) / focal_length_mm
    height_m = (sensor_height_mm * altitude_m) / focal_length_mm
    return width_m, height_m


# Tempo di posa assunto per il calcolo della velocita' massima (ipotesi cautelativa in
# pieno giorno/luce buona: la maggior parte delle camere DJI in modalita' foto sceglie
# automaticamente tempi piu' rapidi di 1/1000s con luce sufficiente). Con poca luce o
# cielo coperto la fotocamera userà tempi più lunghi: in quel caso conviene volare più
# lenti di quanto indicato qui.
ASSUMED_SHUTTER_S = 1 / 1000

# Limite operativo massimo di velocita' di crociera per missioni SAR: oltre i vantaggi
# di sfocatura teorica, un tetto prudenziale che lascia margine per overlap, vento e
# affidabilita' del rilevamento (indicazione diretta, non derivata solo dal motion blur).
MAX_PRACTICAL_SPEED_MS = 7.0
MIN_PRACTICAL_SPEED_MS = 2.0


def recommended_max_speed_ms(gsd_cm_px: float, shutter_s: float = ASSUMED_SHUTTER_S) -> float:
    """Velocita' di avanzamento (m/s) che mantiene lo sfocatura da movimento entro ~1 pixel di GSD.

    Regola pratica di fotogrammetria: a parita' di tempo di posa, piu' il GSD e' fine
    (quota bassa) piu' bisogna volare piano per non sfocare i dettagli; con GSD piu'
    largo (quota alta) si puo' volare piu' veloci senza perdere nitidezza utile.
    """
    gsd_m = gsd_cm_px / 100.0
    speed = gsd_m / shutter_s
    return max(MIN_PRACTICAL_SPEED_MS, min(MAX_PRACTICAL_SPEED_MS, speed))


def recommended_speed_range_ms(
    sensor_width_mm: float,
    focal_length_mm: float,
    image_width_px: int,
    gsd_min_cm: float = GSD_TARGET_MIN_CM,
    gsd_max_cm: float = GSD_TARGET_MAX_CM,
) -> tuple[float, float]:
    """Range di velocita' consigliata (m/s), corrispondente al range di quota consigliato."""
    v_min = recommended_max_speed_ms(gsd_min_cm)
    v_max = recommended_max_speed_ms(gsd_max_cm)
    return v_min, v_max

File: js/mission_planner.js
/**
 * Modulo 1 - Mission Planner Assistant: consigli di volo per la flotta drone in uso.
 * Porta 1:1 la logica di modules/mission_planner.py e pages/1_🛩️_Mission_Planner.py.
 */
window.SF = window.SF || {};

(function () {
  'use strict';

  // ---- Costanti (identiche a modules/mission_planner.py) ----
  const FRONT_OVERLAP_MIN = 70, FRONT_OVERLAP_MAX = 80;
  const SIDE_OVERLAP_MIN = 60, SIDE_OVERLAP_MAX = 70;
  const GSD_TARGET_MIN_CM = 0.8, GSD_TARGET_MAX_CM = 1.5;
  const MAX_PRACTICAL_SPEED_MS = 7.0;

  function altitudeForGsd(gsdCmPx, sensorWidthMm, focalLengthMm, imageWidthPx) {
    return (gsdCmPx * focalLengthMm * imageWidthPx) / (sensorWidthMm * 100);
  }
  function recommendedAltitudeRange(sensorWidthMm, focalLengthMm, imageWidthPx) {
    return [
      altitudeForGsd(GSD_TARGET_MIN_CM, sensorWidthMm, focalLengthMm, imageWidthPx),
      altitudeForGsd(GSD_TARGET_MAX_CM, sensorWidthMm, focalLengthMm, imageWidthPx),
    ];
  }
  function groundFootprintM(altitudeM, sensorWidthMm, sensorHeightMm, focalLengthMm) {
    return [(sensorWidthMm * altitudeM) / focalLengthMm, (sensorHeightMm * altitudeM) / focalLengthMm];
  }

  const mpState = {
    built: false,
    droneId: DRONES[0].id,
    custom: { sensor_width_mm: 6.3, sensor_height_mm: 4.7, image_width_px: 5280, image_height_px: 3956, focal_length_mm: 4.5 },
  };

  function currentSensor() {
    const d = DRONES.find((x) => x.id === mpState.droneId) || DRONES[0];
    const isCustom = d.sensor_width_mm === null;
    if (isCustom) {
      return { ...mpState.custom, drone: d.drone, sensore: d.sensore, note: d.note, isCustom: true };
    }
    return { ...d, isCustom: false };
  }

  function renderResults() {
    const out = document.getElementById('mp-results');
    if (!out) return;
    const sensor = currentSensor();

    if (!sensor.sensor_width_mm || !sensor.focal_length_mm || !sensor.image_width_px) {
      out.innerHTML = '<div class="sf-info">Inserisci i parametri della camera per vedere i consigli di missione.</div>';
      return;
    }
    if (sensor.sensor_width_mm <= 0 || sensor.focal_length_mm <= 0 || sensor.image_width_px <= 0) {
      out.innerHTML = '<div class="sf-error">Parametri non validi: i valori devono essere maggiori di zero.</div>';
      return;
    }

    let thermalNote = '';
    if (!sensor.isCustom && sensor.sensore && sensor.sensore.toLowerCase().includes('termic')) {
      thermalNote =
        '<div class="sf-info">📷 Camera termica: utile per la pianificazione di volo, ma SkyFind individua i target ' +
        'dal colore — usa una camera RGB per la ricerca.</div>';
    }

    const [altMin, altMax] = recommendedAltitudeRange(sensor.sensor_width_mm, sensor.focal_length_mm, sensor.image_width_px);
    const [footW, footH] = groundFootprintM(
      (altMin + altMax) / 2, sensor.sensor_width_mm, sensor.sensor_height_mm, sensor.focal_length_mm
    );
    const maxSpeed = MAX_PRACTICAL_SPEED_MS;

    out.innerHTML = `
      ${thermalNote}
      <h2 class="sf-section">✅ Consigli per questa missione</h2>
      <div class="sf-grid-4">
        <div class="sf-metric"><div class="sf-metric-label">Quota di volo consigliata</div><div class="sf-metric-value">${altMin.toFixed(0)}–${altMax.toFixed(0)} m</div></div>
        <div class="sf-metric"><div class="sf-metric-label">Overlap frontale</div><div class="sf-metric-value">${FRONT_OVERLAP_MIN}–${FRONT_OVERLAP_MAX}%</div></div>
        <div class="sf-metric"><div class="sf-metric-label">Overlap laterale</div><div class="sf-metric-value">${SIDE_OVERLAP_MIN}–${SIDE_OVERLAP_MAX}%</div></div>
        <div class="sf-metric"><div class="sf-metric-label">Velocità massima</div><div class="sf-metric-value">${maxSpeed.toFixed(0)} m/s</div></div>
      </div>
      <div class="sf-success">
        Vola tra <strong>${altMin.toFixed(0)} e ${altMax.toFixed(0)} metri</strong> dal suolo (AGL), con
        <strong>Terrain Following (DSM) attivo</strong> se il terreno non è pianeggiante — mantiene il GSD costante
        ed evita collisioni con dislivelli. A questa quota ogni fotogramma copre circa
        <strong>${footW.toFixed(0)} x ${footH.toFixed(0)} m</strong> a terra — utile per stimare quante strisciate
        servono per coprire l'area di ricerca.
      </div>
      <p><strong>Consigli pratici per lo scatto:</strong></p>
      <ul>
        <li>Imposta l'overlap ai valori indicati sopra nel software di pianificazione missione (es. DJI Pilot 2):
          front overlap ${FRONT_OVERLAP_MIN}-${FRONT_OVERLAP_MAX}%, side overlap ${SIDE_OVERLAP_MIN}-${SIDE_OVERLAP_MAX}%.</li>
        <li>Non superare <strong>${maxSpeed.toFixed(0)} m/s</strong> (${(maxSpeed * 3.6).toFixed(0)} km/h) di velocità di
          crociera: oltre questa soglia rischi foto mosse e overlap insufficiente. Con poca luce, cielo coperto o
          vento vola ancora più piano.</li>
        <li>Inclinazione camera (gimbal): tra <strong>-75° e -85°</strong> (leggermente obliqua) nella maggior parte
          dei casi — aiuta a vedere sotto rami/sporgenze e riduce le ombre nette sul bersaglio. Con vegetazione
          fitta passa a <strong>-90° (nadir puro)</strong>, per massimizzare le possibilità di inquadrare un varco
          nella chioma.</li>
        <li>Imposta il <strong>bilanciamento del bianco fisso</strong> (es. Sole ~5500K o Nuvoloso ~6500K), mai
          automatico: se il drone passa da erba a roccia con l'AWB attivo, il colore dell'indumento nella foto può
          spostarsi abbastanza da uscire dalla tolleranza cromatica calibrata nel modulo 2.</li>
        <li>Mantieni la quota costante durante tutta la missione: variazioni di quota cambiano il GSD e possono
          creare buchi nella copertura.</li>
        <li>Vola con luce diffusa quando possibile: ombre nette rendono più difficile il riconoscimento cromatico
          nel modulo di elaborazione.</li>
      </ul>
      <details class="sf-expander">
        <summary>Dettagli tecnici (sensore e calcolo GSD)</summary>
        <table class="sf-table">
          <tr><th>Parametro</th><th>Valore</th></tr>
          <tr><td>Larghezza sensore</td><td>${sensor.sensor_width_mm} mm</td></tr>
          <tr><td>Altezza sensore</td><td>${sensor.sensor_height_mm} mm</td></tr>
          <tr><td>Focale reale</td><td>${sensor.focal_length_mm} mm</td></tr>
          <tr><td>Risoluzione</td><td>${sensor.image_width_px} x ${sensor.image_height_px} px</td></tr>
        </table>
        ${!sensor.isCustom && sensor.note ? `<p class="sf-caption">ℹ️ ${SF.escapeHtml(sensor.note)}</p>` : ''}
        <p class="sf-caption">Range calcolato per un GSD (dettaglio a terra) target di ${GSD_TARGET_MIN_CM.toFixed(1)}–${GSD_TARGET_MAX_CM.toFixed(1)}
          cm/pixel (verificato sul dataset SAR accademico Heridal, foto scattate tra 40 e 65m, e sulle specifiche
          reali della flotta).</p>
        <p class="sf-caption">Velocità massima: tetto operativo prudenziale di ${maxSpeed.toFixed(0)} m/s (non solo motion
          blur — lascia margine per overlap, vento e affidabilità del rilevamento).</p>
      </details>
    `;
  }

  function renderCustomFields() {
    const wrap = document.getElementById('mp-custom-fields');
    if (!wrap) return;
    const sensor = DRONES.find((x) => x.id === mpState.droneId);
    if (!sensor || sensor.sensor_width_mm !== null) {
      wrap.innerHTML = '';
      return;
    }
    const c = mpState.custom;
    wrap.innerHTML = `
      <p><strong>Il tuo drone non è in elenco: inserisci i parametri della camera manualmente.</strong></p>
      <div class="sf-grid-2">
        <div>
          <label class="sf-label">Larghezza sensore (mm)</label>
          <input type="number" step="0.1" min="0.1" id="mp-sensor-width" value="${c.sensor_width_mm}">
        </div>
        <div>
          <label class="sf-label">Altezza sensore (mm)</label>
          <input type="number" step="0.1" min="0.1" id="mp-sensor-height" value="${c.sensor_height_mm}">
        </div>
        <div>
          <label class="sf-label">Larghezza immagine (px)</label>
          <input type="number" step="1" min="1" id="mp-image-width" value="${c.image_width_px}">
        </div>
        <div>
          <label class="sf-label">Altezza immagine (px)</label>
          <input type="number" step="1" min="1" id="mp-image-height" value="${c.image_height_px}">
        </div>
      </div>
      <label class="sf-label">Lunghezza focale reale (mm, non equivalente)</label>
      <input type="number" step="0.1" min="0.1" id="mp-focal-length" value="${c.focal_length_mm}">
    `;
    const bind = (id, key, isFloat) => {
      document.getElementById(id).addEventListener('input', (e) => {
        const v = isFloat ? parseFloat(e.target.value) : parseInt(e.target.value, 10);
        mpState.custom[key] = Number.isFinite(v) ? v : 0;
        renderResults();
      });
    };
    bind('mp-sensor-width', 'sensor_width_mm', true);
    bind('mp-sensor-height', 'sensor_height_mm', true);
    bind('mp-image-width', 'image_width_px', false);
    bind('mp-image-height', 'image_height_px', false);
    bind('mp-focal-length', 'focal_length_mm', true);
  }

  function buildOnce() {
    const container = document.getElementById('mp-content');
    if (!container) return;
    const options = DRONES.map((d) => `<option value="${d.id}">${SF.escapeHtml(d.drone)} — ${SF.escapeHtml(d.sensore)}</option>`).join('');
    container.innerHTML = `
      <label class="sf-label">Drone e camera in uso</label>
      <select id="mp-drone-select">${options}</select>
      <div id="mp-custom-fields" style="margin-top:0.9rem;"></div>
      <hr style="border-color:var(--border); margin: 1.4rem 0;">
      <div id="mp-results"></div>
    `;
    document.getElementById('mp-drone-select').value = mpState.droneId;
    document.getElementById('mp-drone-select').addEventListener('change', (e) => {
      mpState.droneId = e.target.value;
      renderCustomFields();
      renderResults();
    });
    mpState.built = true;
  }

  SF.init_mission = function () {
    if (!mpState.built) buildOnce();
    renderCustomFields();
    renderResults();
  };
})();
